File: PrimeiraParte/Java/Heranca/exercicio1.java
class Animal {
    private String nome;

    public Animal(String nome) {
        this.nome = nome;
    }

    public String fazerSom() {
        return "Som genérico do animal";
    }

    public String getNome() {
        return nome;
    }
}

class Cachorro extends Animal {
    public Cachorro(String nome) {
        super(nome);
    }

    public String latir() {
        return "Au Au";
    }
}

class Gato extends Animal {
    public Gato(String nome) {
        super(nome);
    }

    public String miar() {
        return "Miau";
    }
}

public class exercicio1 {
    public static void main(String[] args) {
        // Teste
        Cachorro cachorro = new Cachorro("Flokus");
        Gato gato = new Gato("Felicete");

        System.out.println("Cachorro:");
        System.out.println("Nome: " + cachorro.getNome());
        System.out.println("Faz som: " + cachorro.fazerSom());
        System.out.println("Latindo: " + cachorro.latir());

        System.out.println("\nGato:");
        System.out.println("Nome: " + gato.getNome());
        System.out.println("Faz som: " + gato.fazerSom());
        System.out.println("Miau: " + gato.miar());
    }
}

File: PrimeiraParte/C++/Heranca/exercicio1.cpp
#include <iostream>
#include <string>

class Animal {
public:
    Animal(std::string nome) : nome(nome) {}

    std::string fazerSom() {
        return "Som genérico do animal";
    }

    std::string getNome() {
        return nome;
    }

private:
    std::string nome;
};

class Cachorro : public Animal {
public:
    Cachorro(std::string nome) : Animal(nome) {}

    std::string latir() {
        return "Au Au";
    }
};

class Gato : public Animal {
public:
    Gato(std::string nome) : Animal(nome) {}

    std::string miar() {
        return "Miau";
    }
};

int main() {
    // Teste
    Cachorro cachorro("Flokus");
    Gato gato("Felicete");

    std::cout << "Cachorro:\n";
    std::cout << "Nome: " << cachorro.getNome() << "\n";
    std::cout << "Faz som: " << cachorro.fazerSom() << "\n";
    std::cout << "Latindo: " << cachorro.latir() << "\n";

    std::cout << "\nGato:\n";
    std::cout << "Nome: " << gato.getNome() << "\n";
    std::cout << "Faz som: " << gato.fazerSom() << "\n";
    std::cout << "Miau: " << gato.miar() << "\n";

    return 0;
}

